hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -978,3 +997,114 @@ def array_for(index_shape, size):
     return result_shape.flatmap(
         lambda index_shape: st.tuples(*(array_for(index_shape, size) for size in shape))
     )
+
+
+def _from_type(thing: Type[Ex]) -> Optional[st.SearchStrategy[Ex]]:
+    """Called by st.from_type to try to infer a strategy for thing using numpy.
+
+    If we can infer a dtype strategy for thing, we return that; otherwise,
+    returns None (or raises).
+    """
+
+    def unpack_generic(thing):
+        # get_origin and get_args fail on python<3.9 because (some of) the
+        # relevant types do not inherit from _GenericAlias.  So just pick the
+        # value out directly.
+        real_thing = getattr(thing, "__origin__", None)
+        if real_thing is not None:
+            return (real_thing, getattr(thing, "__args__", ()))
+        else:
+            return (thing, ())
+
+    def unpack_dtype(dtype):
+        dtype_args = getattr(dtype, "__args__", ())
+        if dtype_args:
+            assert len(dtype_args) == 1
+            if isinstance(dtype_args[0], TypeVar):
+                # numpy.dtype[+ScalarType]
+                assert dtype_args[0].__bound__ == np.generic
+                dtype = Any
+            else:
+                # plain dtype
+                dtype = dtype_args[0]
+        return dtype
+
+    def find_dtype_shape(args):
+        if len(args) <= 1:
+            # Zero args: ndarray, _SupportsArray
+            # One arg: ndarray[type], _SupportsArray[type]
+            shape = Any
+            dtype = unpack_dtype(args[0]) if args else Any
+        else:
+            # Two args: ndarray[shape, type], NDArray[*]
+            assert len(args) == 2
+            shape = args[0]
+            assert shape is Any
+            dtype = unpack_dtype(args[1])
+        return (
+            scalar_dtypes() if dtype is Any else np.dtype(dtype),
+            array_shapes(max_dims=2) if shape is Any else shape,
+        )
+
+    def base_strats():
+        return [
+            st.booleans(),
+            st.integers(),
+            st.floats(),
+            st.complex_numbers(),
+            st.text(),
+            st.binary(),
+        ]
+
+    if thing == np.dtype:
+        return array_dtypes()
+
+    if thing == ArrayLike:
+        # We override the default type resolution to ensure the "coercible to
+        # array" contract is honoured. See
+        # https://github.com/HypothesisWorks/hypothesis/pull/3670#issuecomment-1578140422
+        base_strat = st.one_of(base_strats())
```

Comment:
I'd probably just assign this instead of having the function above, and then write additional `base_strat_except_bytes` _and_ `base_strat_except_str` (though unfortunately covering both will take yet more code)

Reply:
I'll do that. Actually I was a bit uncertain whether it makes a difference to re-use the same strategy object in multiple places (`one_of(integers(), integers())` vs `i=integers; one_of(i, i)` — I'd think "no", and your comment seems to validate this?

Reply:
Correct, strategies are all cached so reusing vs "re-creating" is in fact almost the same operation (both just fancy dict lookups)